lib/rpc/SingleJsonRpcProvider.ts

```diff
@@ -332,6 +335,8 @@ export class SingleJsonRpcProvider extends StaticJsonRpcProvider {
     } catch (err: any) {
       this.log.error(`Encountered unhandled error when sync provider state: ${JSON.stringify(err)}`)
       // Won't throw. A fail of sync won't affect how we do health state update locally.
+    } finally {
+      this.syncingDb = false
```

Comment:
this is proabbly gonna work, but we keep running into this issue where we have boolean variables that are turned on in one place and you have to turn them on in many other places (similar to "doing helath check" which was your previous PR. Is there a safer way to do this potentially?

Reply:
Since they have to be set at the beginning and the end of an operation (DB sync, evaluate latency, evaluate health), it's hard to combine both access into one.

One thing I could slightly improve the robustness is that,  I will ignore the check for the flag if we haven't recorded the metrics twice long as expected. So even if we made some mistake, we don't deadlock forever. Please see my changes in `checkUnhealthyProviders()` and `checkOtherHealthyProvider()` in `UniJsonRpcProvider.ts`.

Reply:
`if we haven't recorded the metrics twice long as expected` -> what does this mean exactly btw?

Reply:
> `if we haven't recorded the metrics twice long as expected` -> what does this mean exactly btw?

For example, we should wait for the lock if it's already 15 seconds since last healthiness check. However, just in case the lock was failed to release, we will allow a re-acquire of the lock 30 seconds after last healthiness check. This should actually happen, but the logic is here to prevent stalling. Same logic for doing latency check shadow calls.